cmd/installer/pkg/bootloader/syslinux/adv.go

```diff
@@ -0,0 +1,201 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package syslinux
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+const (
+	// AdvSize is the total size.
+	AdvSize = 512
+	// AdvLen is the usable data size.
+	AdvLen = AdvSize - 3*4
+	// AdvMagic1 is the head signature.
+	AdvMagic1 = uint32(0x5a2d2fa5)
+	// AdvMagic2 is the total checksum.
+	AdvMagic2 = uint32(0xa3041767)
+	// AdvMagic3 is the tail signature.
+	AdvMagic3 = uint32(0xdd28bf64)
+)
+
+const (
+	// AdvEnd is the noop tag.
+	AdvEnd = iota
+	// AdvBootonce is the bootonce tag.
+	AdvBootonce
+	// AdvMenusave is the menusave tag.
+	AdvMenusave
+	// AdvReserved1 is a reserved tag.
+	AdvReserved1
+	// AdvReserved2 is a reserved tag.
+	AdvReserved2
+	// AdvReserved3 is a reserved tag.
+	AdvReserved3
+	// AdvUpgrade is the upgrade tag.
+	AdvUpgrade
+)
+
+// ADV represents the Syslinux Auxiliary Data Vector.
+type ADV []byte
+
+// NewADV returns the Auxiliary Data Vector.
+func NewADV(r io.ReadSeeker) (adv ADV, err error) {
+	_, err = r.Seek(-2*AdvSize, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]byte, 2*AdvSize)
+
+	n, err := r.Read(b)
+	if err != nil {
+		return nil, err
+	}
+
+	if n != 2*AdvSize {
+		return nil, fmt.Errorf("expected to read %d bytes, got %d", 2*AdvSize, n)
+	}
+
+	adv = b
+
+	return adv, nil
+}
+
+// ReadTag reads a tag in the ADV.
+func (a ADV) ReadTag(t uint8) (val string, ok bool) {
+	// Header is in first 8 bytes.
+	i := 8
+
+	for {
+		tag := a[i]
+
+		if tag == AdvEnd {
+			break
+		}
+
+		if tag != t {
+			i += 2
+			continue
+		}
+
+		len := int(a[i+1]) + i
+
+		val = string(a[i+2 : len+2])
+
+		ok = true
+
+		break
+	}
+
+	return val, ok
+}
+
+// SetTag sets a tag in the ADV.
+func (a ADV) SetTag(t uint8, val string) (ok bool) {
+	b := []byte(val)
+
+	if len(b) > 255 {
+		return false
+	}
+
+	// Header is in first 8 bytes.
+	i := 8
+
+	for {
+		tag := a[i]
+
+		if tag != AdvEnd {
+			i++
+			continue
+		}
+
+		length := uint8(len(b))
+
+		a[i] = t
+
+		a[i+1] = length
+
+		copy(a[i+2:uint8(i+2)+length], b)
+
+		ok = true
+
+		break
+	}
+
+	if ok {
+		a.cleanup()
+	}
+
+	return ok
+}
+
+// DeleteTag deletes a tag in the ADV.
+func (a ADV) DeleteTag(t uint8) (ok bool) {
+	// Header is in first 8 bytes.
+	i := 8
+
+	for {
+		tag := a[i]
+
+		if tag == AdvEnd {
+			break
+		}
+
+		if tag != t {
+			i++
+			continue
+		}
+
+		b := make([]byte, a[i+1])
+
+		length := uint8(len(b))
+
+		a[i] = 0
```

Comment:
this doesn't seem to be correct if tag being deleted is not the last one, as zeroing out tag area is going to turn it into `AdvEnd`, so tags following it up won't be accessible

instead, we should move the area following the tag N bytes to the left (2 + tag value len)

Reply:
Good catch.